### PR TITLE
ICU-21908 Exact Rounding Increment

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21908
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [ ] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/i18n/unicode/numberformatter.h [hunk 1]
```diff
@@ -640,6 +640,29 @@ class U_I18N_API Precision : public UMemory {
      */
     static IncrementPrecision increment(double roundingIncrement);
 
+#ifndef U_HIDE_DRAFT_API
```
Comment (sffc) [inline icu4c/source/i18n/unicode/numberformatter.h:643]: The API was approved to be:

```
#ifndef U_HIDE_DRAFT_API
    /**
     * Version of `Precision::increment()` that takes an integer at a particular power of 10.
     *
     * To round to the nearest 0.5 and display 2 fraction digits, with this function, you should write one of the following:
     *
     * <pre>
     * Precision::incrementExact(5, -1).withMinFraction(2)
     * Precision::incrementExact(50, -2).withMinFraction(2)
     * Precision::incrementExact(50, -2)
     * </pre>
     *
     * Same as in ICU4J Precision.increment(new BigDecimal(“0.50”)) ...
     *
     * This behavior is modeled after ECMA-402. For more information, see:
     * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#roundingincrement
     *
     * @param mantissa
     *            The increment to which to round numbers.
     * @param magnitude
     *            The power of 10 of the ones digit of roundingIncrement.
     * @return A precision for chaining or passing to the NumberFormatter precision() setter.
     * @draft ICU 71
     */
    static IncrementPrecision incrementExact(uint64_t mantissa, int16_t magnitude);
#endif // U_HIDE_DRAFT_API
```

I need to update the PR with the new signature.

Note that the actual behavior is not changed, so the PR is still reviewable.

File: icu4c/source/i18n/unicode/numberformatter.h [hunk 2]
```diff
@@ -640,6 +640,33 @@ class U_I18N_API Precision : public UMemory {
...
+     *
+     * This is analagous to ICU4J `Precision.increment(new BigDecimal("0.50"))`.
+     *
+     * This behavior is modeled after ECMA-402. For more information, see:
+     * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#roundingincrement
+     *
+     * @param mantissa
+     *            The increment to which to round numbers.
+     * @param magnitude
+     *            The power of 10 of the ones digit of roundingIncrement.
```
Comment (markusicu) [inline icu4c/source/i18n/unicode/numberformatter.h:663]:
```suggestion
     *            The power of 10 of the ones digit of the mantissa.
```

**Review by markusicu — COMMENTED** [review]
looks plausible

File: icu4c/source/i18n/number_rounding.cpp [hunk 3]
```diff
@@ -36,27 +36,26 @@ void number::impl::parseIncrementOption(const StringSegment &segment,
...
+            break;
+        }
     }
-    if (decimalOffset == segment.length()) {
-        outPrecision = Precision::increment(increment);
-    } else {
-        int32_t fractionLength = segment.length() - decimalOffset - 1;
-        outPrecision = Precision::increment(increment).withMinFraction(fractionLength);
+    if (!sawDecimalPoint) {
+        magnitude = dq.adjustToZeroScale();
```
Comment (FrankYFTang) [inline icu4c/source/i18n/number_rounding.cpp:56]: I am lost here. How is magnitude  stored in DecimalQuantity ? I saw you call dq.adjustMagnitude(-magnitude); which flip the magnitude while you pass it to dp, but then in here you took the return value of magnitude from dp w/o flipping it.

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 4]
```diff
@@ -3433,6 +3433,15 @@ void NumberFormatterApiTest::roundingOther() {
...
 
+    assertFormatSingle(
+            u"Large integer increment",
+            u"precision-increment/24000000000000000000000",
+            u"precision-increment/24000000000000000000000",
+            NumberFormatter::with().precision(Precision::incrementExact(24, 21)),
+            Locale::getEnglish(),
+            3.1e22,
+            u"24,000,000,000,000,000,000,000");
+
```
Comment (FrankYFTang) [inline icu4c/source/test/intltest/numbertest_api.cpp:3444]: could you add several more test cases with 1, 5, 10, 25 as increment and ECMA402 limit for maginitute

**Review by FrankYFTang — CHANGES_REQUESTED** [review]
nit on more test cases and thre sign issue

File: icu4c/source/test/intltest/numbertest_api.cpp (hunk 4 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_api.cpp:3444]: fwiw, this feature is already quite well tested and it took work to make all the existing tests pass, but I'll add some more to align with ECMA-402 as you suggested.

File: icu4c/source/i18n/number_rounding.cpp [hunk 5]
```diff
@@ -36,27 +36,26 @@ void number::impl::parseIncrementOption(const StringSegment &segment,
...
-    int decimalOffset = 0;
-    while (decimalOffset < segment.length() && segment.charAt(decimalOffset) != '.') {
-        decimalOffset++;
+    bool sawDecimalPoint = false;
+    int32_t magnitude;
+    for (int32_t i=0; i<buffer.length(); i++) {
+        if (buffer[i] == '.') {
+            sawDecimalPoint = true;
+            magnitude = i - buffer.length() + 1;
+            dq.adjustMagnitude(-magnitude);
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_rounding.cpp:51]: I rearranged the code and added more comments to make it more clear what it is doing.

File: icu4c/source/i18n/number_rounding.cpp (hunk 3 above)
Comment (sffc) [reply icu4c/source/i18n/number_rounding.cpp:56]: I rearranged the code and added more comments to make it more clear what it is doing.

File: icu4c/source/test/intltest/numbertest_api.cpp (hunk 4 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_api.cpp:3444]: Added more tests. Also to Java.

**Comment by markusicu** [comment]
Shane, it sounds like you addressed Frank's feedback but didn't push the commits for your changes.

**Comment by sffc** [comment]
> Shane, it sounds like you addressed Frank's feedback but didn't push the commits for your changes.

What happened is my push failed because I hadn't pulled 35cecbe yet. Fixed.

**Review by FrankYFTang — APPROVED** [review]
LGTM

**Review by FrankYFTang — APPROVED** [review]

**Pull request merged**